### PR TITLE
Add is_closed() method to Queue

### LINKED ISSUE #7355: Add QueueBase.is_closed()
Can we add `bool QueueBase.is_closed()` or `bool QueueBase.is_opened()`?
It will be mainly for debugging purposes.

### PR DESCRIPTION
**Pull request by lattas**
to show if a Queue of any type is closed or not, mainly for debugging purposes. Fixes #7355.

### CODE REVIEW

**Comment by maciekcc** [comment]
Jenkins test this please.

**Comment by a-lattas** [comment]
@girving, you may also be interested to take a look since you discussed this on #7355

File: tensorflow/core/framework/queue_interface.h [hunk 1]
```diff
@@ -77,6 +77,9 @@ class QueueInterface : public ResourceBase {
   virtual void Close(OpKernelContext* ctx, bool cancel_pending_enqueues,
                      DoneCallback callback) = 0;
 
+  // Returns true if a given queue is closed and false if it is open.
+  virtual bool closed() = 0;
```
Comment (girving) [inline tensorflow/core/framework/queue_interface.h]: Should probably be const.

File: tensorflow/core/kernels/queue_ops.cc [hunk 2]
```diff
@@ -425,6 +425,33 @@ class QueueSizeOp : public QueueOpKernel {
 REGISTER_KERNEL_BUILDER(Name("QueueSize").Device(DEVICE_CPU), QueueSizeOp);
 REGISTER_KERNEL_BUILDER(Name("QueueSizeV2").Device(DEVICE_CPU), QueueSizeOp);
 
+// Defines a QueueIsClosedOp, which returns true if a given Queue is closed
+// and false if it is open.
+//
+// The op has one input, which is the handle of the appropriate Queue;
+// and one output, which is a scalar boolean value that shows whether
+// the queue is closed.
```
Comment (girving) [inline tensorflow/core/kernels/queue_ops.cc]: Remove this comment: the REGISTER_OP documentation should cover this.  This kind of comment is in danger of going out of date.

**Review by girving — CHANGES_REQUESTED** [review]

File: tensorflow/core/framework/queue_interface.h (hunk 1 above)
Comment (a-lattas) [inline tensorflow/core/framework/queue_interface.h]: done.

File: tensorflow/core/kernels/queue_ops.cc (hunk 2 above)
Comment (a-lattas) [inline tensorflow/core/kernels/queue_ops.cc]: done.

File: tensorflow/core/framework/queue_interface.h [hunk 3]
```diff
@@ -78,7 +78,7 @@ class QueueInterface : public ResourceBase {
                      DoneCallback callback) = 0;
 
   // Returns true if a given queue is closed and false if it is open.
-  virtual bool closed() = 0;
+  virtual const bool closed() = 0;
```
Comment (girving) [inline tensorflow/core/framework/queue_interface.h]: Should be

    virtual bool closed() const = 0;

Making return values const doesn't do anything in C++.

**Review by girving — CHANGES_REQUESTED** [review]

File: tensorflow/core/framework/queue_interface.h (hunk 3 above)
Comment (a-lattas) [inline tensorflow/core/framework/queue_interface.h]: done.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
@cwhipkey Looks like the tests failed for tool reasons.  Should I ping you or someone else?

**Comment by gunan** [comment]
Both failures seem to be issues on infra side.
I took the bad machines offline. now should be better.

Jenkins, test this please.

**Review by girving — APPROVED** [review]

**Comment by girving** [comment]
@martinwicke Who merges these days?

**Comment by martinwicke** [comment]
This week, @andrewharp.

**Comment by martinwicke** [comment]
Should get API review.

File: tensorflow/core/framework/queue_interface.h [hunk 4]
```diff
@@ -77,6 +77,9 @@ class QueueInterface : public ResourceBase {
   virtual void Close(OpKernelContext* ctx, bool cancel_pending_enqueues,
                      DoneCallback callback) = 0;
 
+  // Returns true if a given queue is closed and false if it is open.
+  virtual bool closed() const = 0;
```
Comment (josh11b) [inline tensorflow/core/framework/queue_interface.h]: For consistency, this should be "is_closed", not "closed"

**Comment by josh11b** [comment]
This is fine for API review assuming that one change is made.

File: tensorflow/core/framework/queue_interface.h (hunk 4 above)
Comment (a-lattas) [inline tensorflow/core/framework/queue_interface.h]: done.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
Looks ready to merge.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by a-lattas** [comment]
The two tests failing are the api-compatibility-test and sparse_ops_test.py.
They have also failed other PRs, as seen in #10671 and #10673 6 hours ago.

**Comment by girving** [comment]
The API test is failing because you changed the API.  Please follow the instructions to make it pass:

    If this test fails, it means a change has been made to the public API. Backwards
    incompatible changes are not allowed. You can run the test as follows to update
    test goldens and package them with your change.

        $ bazel build tensorflow/tools/api/tests:api_compatibility_test
        $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
              --update_goldens True

**Comment by a-lattas** [comment]
@girving Done. Could you please confirm that this is what is needed?

**Review by girving — APPROVED** [review]

**Comment by girving** [comment]
@a-lattas Yep, that's perfect.

Jenkins, test this please.

**Comment by girving** [comment]
@martinwicke The API compatibility test still fails, but it looks a lot like the update script was correctly run.  The update script only changed the `QueueBase` API, though, the errors are for the derived queue classes:

    ERROR:tensorflow:5 differences found between API and golden.
    ERROR:tensorflow:Issue 1	: Change detected in python object: tensorflow.PaddingFIFOQueue.
    ERROR:tensorflow:Issue 2	: Change detected in python object: tensorflow.PriorityQueue.
    ERROR:tensorflow:Issue 3	: Change detected in python object: tensorflow.RandomShuffleQueue.
    ERROR:tensorflow:Issue 4	: Change detected in python object: tensorflow.QueueBase.
    ERROR:tensorflow:Issue 5	: Change detected in python object: tensorflow.FIFOQueue.

Is this a bug in the update script or test?

**Comment by girving** [comment]
Hmm, sorry, I'm wrong: the failure is also for `QueueBase`.  Not sure what's happening then.

**Comment by martinwicke** [comment]
@gunan do you know what could be wrong here?

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Is it possible the PR authors local version is out of sync with the master?
Maybe a pull is needed first?

**Comment by a-lattas** [comment]
@gunan Done.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
@gunan, no luck.

**Comment by a-lattas** [comment]
If I have understood the `api-compatibility-test` correctly, it will fail at any change on the API, except when it is told to update the goldens. Following the instructions in its `README`, I ran the command locally and that should allow only the next `api-compatibility-test` to succeed with API changes, by creating a temporary text file in the same folder. My question is, is the CI aware of this? 

Also note that I added `is_closed` to `Queue_Base` goldens myself. It was not updated automatically. Any thougths?

**Comment by martinwicke** [comment]
The pbtxt files that it generated should be included in the change --
that's how the CI gets to know about it.

**Comment by a-lattas** [comment]
I updated (hardcodded) myself the pbtxt file for QueueBase only. Running 
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
did not update them. Should I update it for all Queue Types or all should be updated automatically?

**Comment by martinwicke** [comment]
If they are not updates that is a sign that there is either something wrong
with the test, or your changes don't actually result in API changes (which
in your case, would be troubling).

**Comment by a-lattas** [comment]
Thank you for the feedback. I will look into it and come up with an answer shortly.

**Comment by gunan** [comment]
I think The problem is still related to syncing.
I checked out your branch, and the test is failing the same way for me.
I think what you needed to do was, first fetch and merge all changes from tensorflow/tensorflow:master. You did this. But you needed to rerun api compatibility test and update goldens again. I can confirm that, when I run api compatibility test on your branch, I see this failure:
```
ERROR:tensorflow:Issue 1	: Change detected in python object: tensorflow.PaddingFIFOQueue.
ERROR:tensorflow:Issue 2	: Change detected in python object: tensorflow.PriorityQueue.
ERROR:tensorflow:Issue 3	: Change detected in python object: tensorflow.RandomShuffleQueue.
ERROR:tensorflow:Issue 4	: Change detected in python object: tensorflow.QueueBase.
ERROR:tensorflow:Issue 5	: Change detected in python object: tensorflow.FIFOQueue.
```
Then I ran update goldens script. Which made the following modifications:
```
	modified:   tensorflow/tools/api/golden/tensorflow.-f-i-f-o-queue.pbtxt
	modified:   tensorflow/tools/api/golden/tensorflow.-padding-f-i-f-o-queue.pbtxt
	modified:   tensorflow/tools/api/golden/tensorflow.-priority-queue.pbtxt
	modified:   tensorflow/tools/api/golden/tensorflow.-queue-base.pbtxt
	modified:   tensorflow/tools/api/golden/tensorflow.-random-shuffle-queue.pbtxt
```
Then rerunning api compatibility test returns success.
```
Ran 2 tests in 1.080s

OK
```

**Comment by drpngx** [comment]
@a-lattas any luck with that?

**Comment by a-lattas** [comment]
I apologise for the delay, I am away from home for a few days.

I can see on @gunan 's post that the `api-compatibility-test` is failing and running it with `update-goldens` let's it progress.

However, when running 
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
```
I get something like
```
Ran 2 tests in 0.080s (skipped 1)
OK
```
with no failure message and 
```
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
does not make any chaninges to the goldens.

I have not found the reason why. `skipped 1` could mean that it is skipping the test? Also, I am not using Docker, as it is not described as necessary. Any ideas?

**Comment by gunan** [comment]
It is certainly possible.
Api compatibility test runs on Linux and python 2 only, as different versions of dependencies can trigger failures in the api object names. Is it possible you are on a macbook, or using python3?

**Comment by a-lattas** [comment]
I suppose this is the issue then. I am using Ubuntu, but with python 3.

I will run the test soon with python 2 and inform you about its result.

Thank you.

**Comment by a-lattas** [comment]
@gunan Thanks! 
I did with python2 . Everything worked as expected and goldens got automatically updated. It should be ready for testing.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
The only failure looks unrelated.  @drpngx: Ready for merge!

**Pull request merged**